sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/RequestResponseChannel.java

```diff
@@ -200,8 +200,6 @@ protected RequestResponseChannel(AmqpConnection amqpConnection, String connectio
                 onTerminalState("SendLinkHandler");
             }),
 
-            //TODO (conniey): Do we need this if we already close the request response nodes when the
-            // connection.closeWork is executed? It would be preferred to get rid of this circular dependency.
```

Comment:
duplicate close of channel is removed in connection.closeWork.

Reply:
Given that the Connection object is responsible for "creating" CBS and Mgmt-nodes objects, I wondered if it logically makes sense for the same Connection object to have the responsibility of "closing" the CBS and Mgmt nodes objects? i.e., the guarantee that by the time parent Connection.close returns, I know the CBS+Mgmt-nodes it owns are cleaned up and closed.

Reply:
Also, the part I'm slightly worried about removing the close-call of CBS + Mgmt-nodes from the Connection is - any edge scenario where the links are dead / reactor-thread go to a bad state, then the hooks in the RRChannel Ctr never get a completion signal hence not call closeAsync (aka Self-Closure) leaving Monos unterminated. But today, at least, the close-call from the Connection ensures the downstream Monos are completed (where we pay for only 2 or 3 extra log lines in this route).

For e.g. thinking about a future (recovery) case where for some reason links are dead, the processor/watchdog closes the Connection but 1 or more authorize calls backed by RRChannel::sendWithAck is waiting forever because the pending Mono never completes..

Reply:
On the other hand, the issue with children (CBS+Mgmt-nodes) observing parent (Connection) closure and reacting to it by self-closing is that the self-closing is async and not chained to the parent async-close. So when the parent async-close returns, technically, the child closure could still be in progress. It's an Ok pattern, but we should avoid it if there is a way to chain it.

Reply:
I've been sketching unique code paths that these two closure-routes (1. close via connection::close and 2. close via shutdownsignal) cover. One case that I think 2 uniquely cover is -


1. thread1 calls conn.getClaimsBasedSecurityNode 
2. Before 1 complete the initialization of cbsChannelProcessor, the connection.closeAsync got called from a thread2.
3. Since cbsChannelProcessor was not completely initialized, connection.closeAsync won't be able to call RRChannel.closeAsync from thread1.
4. The conn.getClaimsBasedSecurityNode completes the initialization.
5. Since the connection set shutdownsignal before its closure, the RRChannel will be able to react to it and do self-cleanup.

So although the two closure route are defined & may overlap, they seem to cover unique cases in the graph. So we might have to take a step back and think, is it ok to have those 3 or 4 lines of extra logging if that guarantees reliability JM2C.

Reply:
>The remoteClose never gets fired but always timeout

After debugging, I found CBS channel didn't send out the `Detach` frame when closing, and it only sent the `End` frame. Currently I think it may because CBS `session.close()` is called earlier than `link.close()`. Inside Proton-J, When `TransportImpl#processEnd()` is triggered, it would call `freeLocalChannel()` to clear up links' `_localHandle` and then send out the `End` frame. As links' `_localHandle` is removed, the `Detach` frame won't send out even we manually call the `link.close()`.
```
14:27:42.286 [main] DEBUG c.a.c.a.implementation.ReactorSender - {"az.sdk.message":"Setting error condition and disposing. Connection shutdown.","connectionId":"MF_77c468_1648708050569","entityPath":"eventhub-test","linkName":"eventhub-test","errorCondition":"n/a","errorDescription":"n/a"}
14:27:43.859 [reactor-executor-1] DEBUG c.a.c.a.i.ReactorSession - {"az.sdk.message":"Got endpoint state.","connectionId":"MF_77c468_1648708050569","sessionName":"eventhub-test","state":"CLOSED"}
14:27:43.860 [reactor-executor-1] DEBUG c.a.c.a.i.ReactorSession - {"az.sdk.message":"Disposing of active send and receive links due to session close.","connectionId":"MF_77c468_1648708050569","sessionName":"eventhub-test"}
14:27:43.860 [reactor-executor-1] DEBUG c.a.c.a.i.ReactorSession - {"az.sdk.message":"Got endpoint state.","connectionId":"MF_77c468_1648708050569","sessionName":"cbs-session","state":"CLOSED"}
14:27:43.860 [reactor-executor-1] DEBUG c.a.c.a.i.ReactorSession - {"az.sdk.message":"Disposing of active send and receive links due to session close.","connectionId":"MF_77c468_1648708050569","sessionName":"cbs-session"}
14:27:43.861 [reactor-executor-1] DEBUG c.a.c.a.i.handler.SessionHandler - {"az.sdk.message":"onSessionLocalClose","connectionId":"MF_77c468_1648708050569","errorCondition":null,"errorDescription":null,"sessionName":"eventhub-test"}
14:28:39.294 [main] DEBUG c.a.c.a.i.RequestResponseChannel - {"az.sdk.message":"Closing request/response channel.","connectionId":"MF_77c468_1648708050569","linkName":"cbs"}
14:29:07.952 [reactor-executor-1] DEBUG c.a.c.a.i.handler.SessionHandler - {"az.sdk.message":"onSessionLocalClose","connectionId":"MF_77c468_1648708050569","errorCondition":null,"errorDescription":null,"sessionName":"cbs-session"}
14:29:07.952 [reactor-executor-1] DEBUG c.a.c.a.i.handler.SendLinkHandler - {"az.sdk.message":"onLinkLocalClose","connectionId":"MF_77c468_1648708050569","errorCondition":null,"errorDescription":null,"linkName":"eventhub-test"}
14:29:07.956 [reactor-executor-1] DEBUG c.a.c.a.i.RequestResponseChannel - {"az.sdk.message":"Closing send link and receive link.","connectionId":"MF_77c468_1648708050569","linkName":"cbs"}
```
I will try to summarize this part of closing logic and may find out a way to close cbs links before session closed.

Reply:
Because session is also created by connection, currently I add a `closeSessionOperation` inside `connection.closeAsync()` and connection closing logic becomes:
1. close CBS + Mgmt node
2. close session 
3. emit shutdown signal to close sender/receiver/session/RRchannel 
4. close connection
```Java
final Mono<Void> sessionCloseOperation = Mono.when(
            Flux.fromStream(sessionMap.values().stream())
            .flatMap(sessionSubscription -> sessionSubscription.getSession().closeAsync()));

return Mono.whenDelayError(
    cbsCloseOperation.doFinally(signalType ->
         ...
    managementNodeCloseOperations.doFinally(signalType ->
        ...
    .then(sessionCloseOperation.doFinally(signalType ->
        ...
    .then(emitShutDownSignalOperation.doFinally(signalType ->
        ...
    .then(closeReactor.doFinally(signalType ->
        ...
    .then(isClosedMono.asMono());
```
By this change, we could send out the `Detach` frame for CBS links and successfully trigger remoteClose. I will double check the session closing logic and update code.

Reply:
Thanks, Kun, for digging into this. It's good to see that the Session.END frame was the one overriding the sending of Link(s):DEATCH. It is expected behavior as per amqp-protocol; the Session close will be considered an "implicit" closing of the links from both server and client-side. Less traffic and there is no handle-leak given its a valid.

Reply:
Also, it would be nice to add a comment to the RRChannel type (this help us/future-devs) where we're listening for connection ShutdownSignals and doing a self-close i.e., it's "_to ensure graceful closure of RRChannel instance that won the race between its creation and its parent connection close._"

```java
amqpConnection.getShutdownSignals().next()....
```

Reply:
Sure. I will add more comments and update docs base on the latest logic.